File: 3Sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums = sorted(nums)
        for i in range(0,len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            target = 0 - nums[i]
            left = i + 1
            right = len(nums) - 1
            while left < right:
                if nums[left] + nums[right] == target:
                    res.append([nums[i],nums[left],nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
                elif nums[left] + nums[right] < target:
                    left += 1
                else:
                    right -= 1
        return res

File: 3Sum/3sum.go
func threeSum(nums []int) [][]int {
    sort.Ints(nums)
    var res [][]int
    for i := 0; i < len(nums) - 2; i++ {
        if i > 0 && nums[i - 1] == nums[i] {
            continue
        }
        var target int = 0 - nums[i]
        var left int = i + 1
        var right int = len(nums) - 1
        for left < right {
            if nums[left] + nums[right] == target {
                res = append(res, []int{nums[i],nums[left],nums[right]})
                for left < right && nums[left] == nums[left + 1] {left++}
                for left < right && nums[right] == nums[right - 1] {right--}
                left++
                right--
            }else if nums[left] + nums[right] < target {
                left++
            }else {
                right--
            }
        }
    }
    return res;
}
